proposals/001-operator-design.md

```diff
@@ -0,0 +1,818 @@
+  cluster we need that label to have additional structure (naming convention). E.g. foo-broker-0, foo-broker-1 etc. We
+  need the ‘foo’ part because we'd like to support multiple clusters in the same namespace (so bar-broker-0,
+  bar-broker-1). But applying additional structure to the DNS label like this is not compatible with TLS wildcard
+  certificates. So we end with a certificate provisioning problem (e.g. Kafka cluster scale up => Need more Services =>
+  Need new certificates).
+* It follows from this that the hostname property on Gateway is kind of useless for Kafka.
+    * For a non-wildcard hostname it must identify a single broker, which is useless.
+    * For a wildcard hostname in the cluster.local domain it must identify all the brokers (from multiple Kafka
+      clusters) in that namespace.
+
+#### A possible solution
+
+* Don’t support hostname on KafkaGateway at all (which is compatible because it’s optional in Gateway)
+* Support hostname on KafkaRoute, but don’t support traditional wildcards (using *, which can match multiple DNS labels)
+* Instead, support a different wildcard syntax, using %, which matches only a part of a single DNS label (require it to
+  be in the first label and to have some prefixing characters)
+
+In the Gateway API you can declare a list of hostnames for a route. In the proxy we have two sets of hostnames to route
+to a Virtual Cluster; a bootstrap SNI hostname, and a brokerAddress pattern which routes a pattern to the right cluster.
+We would like each hostname entry in the list to fully describe a virtual cluster. So we introduce a hostname scheme
+convention:
+
+1. Bootstrap address = ${prefix}bootstrap.${suffix}
+2. Broker address for broker id N = ${prefix}broker-${N}.${suffix}
+
+By requiring a prefix we can validate that the hostnames are disjoint and we can unambiguously route traffic for a
+hostname to a Route.
+
+This means the user will specify a hostname with notation like my-cluster-%.kafka.example.com and, following our
+convention, we will determine that the bootstrap hostname is my-cluster-boostrap.kafka.example.com and the node id
+pattern is my-cluster-broker-$(nodeId).kafka.example.com.
+
+We don’t anticipate users needing great control over the first hostname section.
+
+Example KafkaRoute spec
+
+```yaml
+---
+kind: KafkaRoute
+apiVersion: kroxylicious.io/v1alpha1
+metadata:
+  name: my-route
+spec:
+  parentRefs:
+    - # A reference to the KafkaProxies this Route is a part of
+      group: kroxylicious.io # Must be kroxylicious.io
+      kind: KafkaGateway # Must be KafkaGateway
+      name: simple
+      sectionName: my-listener
+      brokers:
+        - advertisedBrokerIds: [ 1,2,3,4,5,6,7,8,9 ]
+          brokerIdentityMapping: identity | hash # to be added eventually, default is identity
+      hostnames:
+        # When a hostnamePattern ends with the cluster/namespace 
```

Comment:
We are using a naming convention to let the user choose cluster ip or load balancer (or both).   How would we gesture that we want to use an alternative off-cluster ingress mechanism?  Say, a Gateway or an OpenShift Route?